### ISSUE TITLE
Refactor recommender to "CAS in - CAS out"

### ISSUE THREAD

**Issue by Rentier** [inception-project/inception#398]
**Describe the refactoring action**
Refactor the recommender so they use `CAS` both as input and output.

**Expected benefit**
It will be easier to test, cleaner, looser coupled and will allow us to implement recommender that were not possible before, e.g. relation recommender.

**Things to do in PR #399**

- [x] Adjust Prediction and selection tasks (@Rentier PR #399)
- [x] Port and test DL4J (@reckart PR #399)
- [x] Rewrite, port and test String based recommender (@reckart PR #399)
- [x] Port and test OpenNLP POS/NER recommenders (@Rentier, @reckart PR #399)
- [x] Port and test External recommender (@Rentier,  PR #399) 
- [x] Port and test Named entity linker (@rcffc, PR #399) 
- [x] Reintroduce traits

**Things to do in follow-up PR**

- [ ] Refactor annotation object
- [ ] Port active learning
- [ ] Parameterized recommender smoke test
- [ ] Port renderer (with lots of unit tests)
- [ ] JavaDoc
- [ ] Developer documentation
- [ ] Tutorial to implement simple recommender
- [ ] Adhere to max suggestions

**Comment by reckart** [comment]
I believe it is not necessary to use concurrent collections in the `RecommenderContext` because it once it has been created, it should only be shared read-only.

**Comment by reckart** [comment]
I have added a JCas class called `PredictedSpan` to the recommender API module.

**Comment by reckart** [comment]
I have reactivated the DL4J module again and ported it to the new API. It will still require some work to be really usable, but in general it seems to work ok with the new API.

Originally, I wanted to keep the data I keep in memory small by processing CASes one-by-one. However, I ended up extract all the training data from all CASes before starting the training process because otherwise, I couldn't properly split it into training and test sets.

The unit tests you did for the OpenNlpNer could be ported to the DL4J recommender almost copy/paste. That was also pretty cool. Maybe we can later set up some parametrized test suite which can be applied to multiple recommenders instead of maintaining per-recommender tests.

**Comment by reckart** [comment]
@rcffc @Rentier I have organized the todo list here at the top. Has anybody already started working on re-introducing traits and/or porting the OpenNLP recommenders?

**Comment by Rentier** [comment]
I thought that POS (you) and NER (me) is already done.

**Comment by reckart** [comment]
@Rentier I thought that as well, but the checkbox wasn't ticked yet - ok, I ticked it and put our names there.

**Comment by rcffc** [comment]
Can we add a description field to PredictedSpan?

**Comment by reckart** [comment]
> Can we add a description field to PredictedSpan?

We probably could, but I am not sure if we should. I assume you ask in order to be able to display a description obtained from the KB for a linked entity. I think instead of adding this to the prediction, the description should be obtained during rendering - or even better - on demand when/if the user actually opens the hover-popup.

@Rentier WDYT?

**Comment by Rentier** [comment]
This will be done when we refactor the AnnotationObject and SpanRenderer. The PredictedSpan just lives inside recommenders and never gets in contact with the UI.

**Comment by reckart** [comment]
@Rentier I take that as **descriptions are not added to the predicted span because descriptions are not predicted by the recommender but rather are auxiliary information that is picked up during rendering**.

**Comment by Rentier** [comment]
It looks like the features are complete, imho we can start to test and review this beast.

**Comment by Rentier** [comment]
I would honestly disable the dl4j, I cannot run these tests even with 4G memory, they are really slow and I doubt the usefulness of this recommender (even more data hungry and slow compared to OpenNLP).

**Comment by reckart** [comment]
@Rentier how about moving the remaining todos into a new issue and closing this one?
